### PR TITLE
feat(info): adding more functionality to info package

### PR DESCRIPTION
**Pull request by pranshuchittora**
Today, while working on this [PR](https://github.com/webpack/webpack-cli/pull/832), I observed that many CLI `packages` have very fewer downloads.

 `info` package downloads  [npm](https://www.npmjs.com/package/@webpack-cli/info)

As a developer, I have expected the `info` package to give some information about config or something related to webpack. Strangely it outputs some environment information.

So the idea is to revamp the current `info` package by adding a few features.

- [x]  Information about webpack config. By specifying the path of the webpack config, it will provide some information about plugins, entry point, output point...

Screenshot from 2019-04-18 02-53-53
_Expected behaviour_

Solution: By using the power of AST we can parse the config file & get the required details.

- [x] Extending the current environment information.
The current env info looks good, but someone might want some other env info, so the idea is to add more env info & access them dynamically by flags.

Screenshot from 2019-04-18 02-54-46

Screenshot from 2019-04-18 02-55-24
_Multiple flags supported, in order to get multiple env info_

Solution: By reading the process arguments (flags)

**What kind of change does this PR introduce?**
Feature
**Did you add tests for your changes?**
Yes
**If relevant, did you update the documentation?**
Yes

I am working on this, so pls suggest any improvements. 🙏
Let's discuss :fire:

### CODE REVIEW

**Comment by ematipico** [comment]
Now that you're proposing some new flags, which are Great, we should also provide an help command/flag so we show how to use the `info`

**Comment by pranshuchittora** [comment]
@ematipico check this out.

- [x] Help flag

Screenshot from 2019-04-18 18-05-32

### Output formats
- [x] JSON

Screenshot from 2019-04-18 18-11-43

- [x] Markdown

Screenshot from 2019-04-18 18-12-31
Markdown format would make issues to look more cleaner.

Default

Screenshot from 2019-04-18 18-14-40

**Comment by anikethsaha** [comment]
This one is really great . You can give the command name  like `describe configfilename` much like in kubernetes does to output the state and configuration of the file.

**Comment by pranshuchittora** [comment]
@evenstensberg  I am thinking of going with ASTs to extract the required information from the config file. Is there any other way around :sunglasses:

File: packages/info/index.ts [hunk 1]
```diff
@@ -1,20 +1,38 @@
...
 
-export function information() {
-	return {
-		Binaries: ["Node", "Yarn", "npm"],
-		Browsers: ["Chrome", "Firefox", "Safari"],
-		System: ["OS", "CPU"],
-		npmGlobalPackages: ["webpack", "webpack-cli"],
-		npmPackages: "*webpack*",
-	};
+const AVAILABLE_FLAGS: string[] = ["binaries", "system", "browsers", "npmGlobalPackages", "npmPackages"];
```
Comment (ematipico) [inline packages/info/index.ts]: These strings should be constants! Write them once and change them once! :)

**Comment by ematipico** [comment]
@evenstensberg why did you chill a PR that is still WIP?

**Comment by pranshuchittora** [comment]
Progress so far
- `webpack-cli info ` _(Default)_
Screenshot from 2019-04-21 18-41-26
- `webpack-cli info --json`
Screenshot from 2019-04-21 18-42-46

- `webpack-cli info --markdown`
Screenshot from 2019-04-21 18-43-22

- `webpack-cli info --sys --bin --json`
Screenshot from 2019-04-21 18-37-52

- `webpack-cli info --sys --bin --json --xyz`
Screenshot from 2019-04-21 18-38-50
- `webpack-cli info --help`
Screenshot from 2019-04-21 18-39-31

- `webpack-cli info --www --xyz --bin --npmg --markdown`
Screenshot from 2019-04-21 18-44-13

- `webpack-cli info --system --binaries --browsers --markdown`
Screenshot from 2019-04-21 18-40-28

**Comment by pranshuchittora** [comment]
:1st_place_medal:  Milestone 1

- [x] Support for flags to make it dynamic 

- [x] Support for JSON and Markdown

- [x] Support for aliases

- [x] Add `--help` flag

- [ ] Update tests

- [ ] Update docs

- [x] Logging on invalid options

File: junit.xml [hunk 2]
```diff
@@ -1,557 +1,565 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (misterdev) [inline junit.xml]: Is this file needed? :)

Comment (pranshuchittora) [reply junit.xml]: Yep, this file is associated with tests.

File: packages/info/index.ts [hunk 3]
```diff
@@ -1,22 +1,57 @@
...
+		} else {
+			// Invalid option
+			process.stdout.write("\n" + chalk.bgRed(flag) + chalk.red(" is an invalid option"));
+			return;
+
+		}
+	});
+
+	process.stdout.write("\n" + await envinfo.run(
+		Object.keys(DETAILS_OBJ).length ?
```
Comment (evenstensberg) [inline packages/info/index.ts]: Extract this into a variable

File: packages/info/index.ts [hunk 4]
```diff
@@ -1,22 +1,57 @@
...
+			return;
+		} else if (AVAILABLE_COMMANDS.includes(flag)) {
+			const flagVal = informationType(flag);
+			DETAILS_OBJ = { ...DETAILS_OBJ, ...flagVal };
+		} else if (AVAILABLE_FORMATS.includes(flag)) {
+			CONFIG[flag] = true;
+		} else {
+			// Invalid option
+			process.stdout.write("\n" + chalk.bgRed(flag) + chalk.red(" is an invalid option"));
+			return;
```
Comment (rishabh3112) [inline packages/info/index.ts]: return here is unneeded as no more statement are left in callback. You instead exit the process at this point after printing.

File: packages/info/index.ts [hunk 5]
```diff
@@ -1,22 +1,57 @@
...
+		case "npm":
+			return { npmPackages: "*webpack*" };
+
+	}
 }
 
 export default async function info(): Promise<void> {
-	process.stdout.write(await envinfo.run(information()));
+
+	Object.keys(argv).forEach(async (flag): Promise<void> => {
```
[the remark this answers is not in the document]
Comment (pranshuchittora) [reply packages/info/index.ts]: https://github.com/webpack/webpack-cli/pull/833/files#diff-b8cb449e59f8e92900349cd6fcf144b2R52

File: packages/info/index.ts (hunk 4 above)
Comment (pranshuchittora) [reply packages/info/index.ts]: The `forEach` iterator iterates over the given array and after that the remaining instructions in the `info()` func gets executed.

**Comment by pranshuchittora** [comment]
> Hi, nice work. Could you add description of the API in the readme? After that I'll do a final check

@evenstensberg  Done.

What about bumping the package version

File: packages/info/index.ts (hunk 4 above)
Comment (pranshuchittora) [reply packages/info/index.ts]: As the func is going to be imported, hence proper returning is required.
> Have a look at the new update

Comment (rishabh3112) [reply packages/info/index.ts]: Returning from callback is not going to return from info function, just from callback. Ending process here is better IMO.

**Comment by evenstensberg** [comment]
yes, leave that to me. Could you rebase?

**Comment by pranshuchittora** [comment]
Any suggestions regarding, parsing the config?

**Comment by evenstensberg** [comment]
take in a new param, path resolve

**Comment by pranshuchittora** [comment]
I meant reading the different properties from the config, which consists of different data types.

**Comment by evenstensberg** [comment]
You traverse the config and check types, thats not needed. Output metadata, entry, output and things like that. Simple stuff dont make it too complicated

**Comment by pranshuchittora** [comment]
Screenshot from 2019-06-10 23-19-03

File: packages/info/README.md [hunk 6]
```diff
@@ -9,22 +9,55 @@ This package returns a set of information related to the local environment.
...
+yarn add @webpack-cli/info -D
 ```
 
 ## Usage
 
+### Args / Flags
+
+#### Output format
+
+| Flag         | Description                   | Type        |     |     |
```
Comment (ematipico) [inline packages/info/README.md]: Output flags should be `--output-json`, `--output-markdown`

**Comment by pranshuchittora** [comment]
Based on new init (v3.3.4)

Screenshot from 2019-06-11 14-43-41
Screenshot from 2019-06-11 14-44-06

File: packages/info/README.md [hunk 7]
```diff
@@ -9,22 +9,55 @@ This package returns a set of information related to the local environment.
...
+| `--json`     | To get the output as JSON     | [ boolean ] |     |     |
+| `--markdown` | To get the output as markdown | [ boolean ] |     |     |
+
+#### Options
+
+| Flag                   | Description                                                    | Type        |
+| ---------------------- | -------------------------------------------------------------- | ----------- |
+| `--help`               | Show help                                                      | [ boolean ] |
+| `--version`            | Show version number                                            | [ boolean ] |
+| `--system` , `--sys`   | System information ( OS, CPU )                                 | [ boolean ] |
```
Comment (ematipico) [inline packages/info/README.md]: Aliases should be just one letter

File: packages/info/README.md [hunk 8]
```diff
@@ -9,22 +9,55 @@ This package returns a set of information related to the local environment.
...
+| ------------ | ----------------------------- | ----------- | --- | --- |
+| `--json`     | To get the output as JSON     | [ boolean ] |     |     |
+| `--markdown` | To get the output as markdown | [ boolean ] |     |     |
+
+#### Options
+
+| Flag                   | Description                                                    | Type        |
+| ---------------------- | -------------------------------------------------------------- | ----------- |
+| `--help`               | Show help                                                      | [ boolean ] |
+| `--version`            | Show version number                                            | [ boolean ] |
```
Comment (ematipico) [inline packages/info/README.md]: Version number of what? We need to be specific in the description

File: packages/info/README.md [hunk 9]
```diff
@@ -9,22 +9,55 @@ This package returns a set of information related to the local environment.
...
+| `--system` , `--sys`   | System information ( OS, CPU )                                 | [ boolean ] |
+| `--binaries` , `--bin` | Installed binaries                                             | [ boolean ] |
+| `--browsers`           | Installed web browsers                                         | [ boolean ] |
+| `--npmg`               | Globally installed NPM packages ( webpack & webpack-cli only ) | [ boolean ] |
+| `--npmPackages`        | Info about webpack installed in the project                    | [ boolean ] |
```
Comment (ematipico) [inline packages/info/README.md]: The description is not good. I read it and I don't understand what's the command is about for

File: packages/info/README.md [hunk 10]
```diff
@@ -9,22 +9,55 @@ This package returns a set of information related to the local environment.
...
+
 ### Node
 
 ```js
-const envinfo = require("@webpack-cli/info").default;
-envinfo();
+const info = require("webpack-cli/packages/info").default;
```
Comment (ematipico) [inline packages/info/README.md]:
```suggestion
const info = require("@webpack-cli/info").default;
```

File: packages/info/README.md [hunk 11]
```diff
@@ -9,22 +9,55 @@ This package returns a set of information related to the local environment.
...
+
+async function wrapperFunc() {
+	let envinfo = await info({
```
Comment (ematipico) [inline packages/info/README.md]: Is the config needed?

Comment (ematipico) [inline packages/info/README.md]: `envinfo` it's not used, it can be removed

File: packages/info/index.ts [hunk 12]
```diff
@@ -1,29 +1,77 @@
...
+			return { Browsers: ["Chrome", "Firefox", "Safari"] };
+		case "npmg":
+			return { npmGlobalPackages: ["webpack", "webpack-cli"] };
+		case "npm":
+			return { npmPackages: "*webpack*" };
+	}
 }
+export default async function info(CustomArgv: object): Promise<void> {
+	const CUSTOM_AGRUMENTS: boolean = typeof CustomArgv === "object";
+	const args: ArgvI = CUSTOM_AGRUMENTS ? CustomArgv : argv;
```
Comment (ematipico) [inline packages/info/index.ts]: Why don't you use `yargs` to parse the arguments?

File: packages/info/README.md (hunk 11 above)
Comment (evenstensberg) [reply packages/info/README.md]: Supplying a config from `--config` is great I think

File: packages/info/configParser.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
+import * as path from "path";
+import chalk from "chalk";
+import * as prettyjson from "prettyjson";
+
+export function getNameFromPath(fullPath: string): string {
+	const filename = fullPath.replace(/^.*[\\\/]/, "");
+	return filename;
+}
+
+export function resolveFilePath(relativeFilePath: string) {
```
Comment (sendilkumarn) [inline packages/info/configParser.ts]: return type? here and below?

File: packages/info/configParser.ts [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+	const configPath = path.resolve(process.cwd() + "/" + relativeFilePath);
+	return configPath;
+}
+
+export function fetchConfig(configPath: string) {
+	let config = null;
+	config = require(configPath);
+	try {
```
Comment (sendilkumarn) [inline packages/info/configParser.ts]: what is happening here? why empty try? Do you mean to put config inside the try block?

File: packages/info/configParser.ts [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+	} catch (e) {
+		process.stdout.write(chalk.red(`Error:`, e.code) + "\n");
```
Comment (sendilkumarn) [inline packages/info/configParser.ts]: can you stick either with the template string or normal string concatenation rather than mixing them both?

File: packages/info/configParser.ts [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+	}
+	return config;
+}
+
+export function configReader(config) {
```
Comment (sendilkumarn) [inline packages/info/configParser.ts]: I would prefer to have `type` information here as well

File: packages/info/README.md (hunk 11 above)
Comment (ematipico) [reply packages/info/README.md]: I meant the variable `envinfo`. Is it needed in this example?

File: packages/info/README.md (hunk 6 above)
Comment (pranshuchittora) [reply packages/info/README.md]: Sounds great :+1: 
Done :rocket:

File: packages/info/README.md (hunk 7 above)
Comment (pranshuchittora) [reply packages/info/README.md]: Done :+1:

File: packages/info/README.md (hunk 8 above)
Comment (pranshuchittora) [reply packages/info/README.md]: The version of `webpack-cli`
Just updated the docs :+1:

File: packages/info/README.md (hunk 11 above)
Comment (pranshuchittora) [reply packages/info/README.md]: @evenstensberg shall I change the config input to  `--config=[PATH]` a flag?

File: packages/info/index.ts (hunk 12 above)
Comment (pranshuchittora) [reply packages/info/index.ts]: Pls elaborate.

File: packages/info/configParser.ts (hunk 14 above)
Comment (pranshuchittora) [reply packages/info/configParser.ts]: Good catch, actually while debugging I shifted the require statement out of the `try`.

After making require the child of `try`. (When module related err occurs)
Screenshot from 2019-06-12 00-46-39

File: packages/info/configParser.ts (hunk 13 above)
Comment (pranshuchittora) [reply packages/info/configParser.ts]: Done

File: packages/info/configParser.ts (hunk 15 above)
Comment (pranshuchittora) [reply packages/info/configParser.ts]: Ok

File: packages/info/index.ts (hunk 12 above)
Comment (ematipico) [reply packages/info/index.ts]: `yargs` should be used to parse the arguments passed to the cli. That's what we have. We should also install the `@types/yargs` if we didn't, so we already have the correct interfaces

File: packages/info/README.md [hunk 17]
```diff
@@ -9,22 +9,73 @@ This package returns a set of information related to the local environment.
...
+wrapperFunc();
 ```
 
+#### Custom config
+
+> Config has higher precedence than system flags
+
+```json
+{
+    // Config relative path
```
Comment (ematipico) [inline packages/info/README.md]: JSON format can't have comments. If you want have comments, could you change format?

File: packages/info/commands.ts [hunk 18]
```diff
@@ -0,0 +1,13 @@
+export const AVAILABLE_COMMANDS: string[] = [
```
Comment (ematipico) [inline packages/info/commands.ts]: You use these constants inside your commands.ts, so you can keep your code DRY.

File: packages/info/index.ts (hunk 12 above)
Comment (ematipico) [reply packages/info/index.ts]: You define the options using yargs, so could you please use yargs to parse the arguments too? It will keep consistency. Sorry to be annoying but I don't see any reason to not do that

File: packages/info/README.md [hunk 19]
```diff
@@ -9,22 +9,73 @@ This package returns a set of information related to the local environment.
...
 
 ```bash
+#npm
 npm i -D @webpack-cli/info
+
+#yarn
+yarn add @webpack-cli/info -D
 ```
 
 ## Usage
```
Comment (rishabh3112) [inline packages/info/README.md]: Add up usage command here
```
npx webpack info [options]
```

File: packages/info/README.md (hunk 17 above)
Comment (pranshuchittora) [reply packages/info/README.md]: I know JSON doesn't have comments, it for demonstration purpose only. Will change the format.

File: packages/info/commands.ts (hunk 18 above)
Comment (pranshuchittora) [reply packages/info/commands.ts]: I am not getting it.

File: packages/info/index.ts (hunk 12 above)
Comment (pranshuchittora) [reply packages/info/index.ts]: I am using yargs to parse the arguments. Though I am not using `yargs-parser`.

File: packages/info/commands.ts (hunk 18 above)
Comment (ematipico) [reply packages/info/commands.ts]: 1. You should create an `enum` for your available commands.
2. Later you will use AVAILABLE_COMMANDS.BINARIES inside `commands.ts`

Basically it's not convenient to have the same string repeated in different places. Create ONE constant and it in those places

File: packages/info/README.md [hunk 20]
```diff
@@ -9,22 +9,79 @@ This package returns a set of information related to the local environment.
...
 
 ```bash
+#npm
 npm i -D @webpack-cli/info
+
+#yarn
+yarn add @webpack-cli/info -D
+
+#npx
+npx webpack info [options]
```
Comment (rishabh3112) [inline packages/info/README.md]: It's not an installation command, it should be under usage :)

File: packages/info/index.ts (hunk 12 above)
Comment (ematipico) [reply packages/info/index.ts]: Why not?

**Comment by evenstensberg** [comment]
Could you rebase this against the next branch?

**Comment by pranshuchittora** [comment]
Refer #1007
